homeassistant/components/rainbird/__init__.py

```diff
@@ -3,40 +3,79 @@
 
 import voluptuous as vol
 
+from homeassistant.components import binary_sensor, sensor, switch
+from homeassistant.const import (
+    CONF_FRIENDLY_NAME,
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_TRIGGER_TIME,
+)
+from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
-from homeassistant.const import CONF_HOST, CONF_PASSWORD
 
 _LOGGER = logging.getLogger(__name__)
 
+RAINBIRD_CONTROLLER = "controller"
 DATA_RAINBIRD = "rainbird"
 DOMAIN = "rainbird"
 
-CONFIG_SCHEMA = vol.Schema(
+SENSOR_TYPE_RAINDELAY = "raindelay"
+SENSOR_TYPE_RAINSENSOR = "rainsensor"
+# sensor_type [ description, unit, icon ]
+SENSOR_TYPES = {
+    SENSOR_TYPE_RAINSENSOR: ["Rainsensor", None, "mdi:water"],
+    SENSOR_TYPE_RAINDELAY: ["Raindelay", None, "mdi:water-off"],
+}
+
+ZONE_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {vol.Required(CONF_HOST): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-        )
-    },
+        vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+        vol.Optional(CONF_TRIGGER_TIME): cv.positive_int,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.positive_int,
+    }
+)
+CONTROLLER_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_TRIGGER_TIME): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.string,
+        vol.Optional("zones"): vol.Schema({cv.positive_int: ZONE_SCHEMA}),
+    }
+)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.Any({cv.string: CONTROLLER_SCHEMA}, CONTROLLER_SCHEMA))},
```

Comment:
The controller id doesn't need to be a config item. We can just use some arbitrary integer as index. Use `cv.ensure_list` and make it a list of controller schemas.
```py
CONFIG_SCHEMA = vol.Schema(
    {DOMAIN: vol.Schema(vol.All(cv.ensure_list, [CONTROLLER_SCHEMA]))},

Reply:
refactored